### ISSUE TITLE
No '_get_object_id' while running a checkpoint using a `RuntimeBatchRequest`

### ISSUE THREAD

**Issue by tnorlund-dsc** [fivetran/great_expectations#3473]
**Describe the bug**
I get an error while attempting to run a checkpoint using a `RuntimeBatchRequest`.

**To Reproduce**
# Import
```
import sys
from datetime import datetime
from pyspark.sql.functions import *
from pyspark.sql.types import StructType, StructField, StringType, IntegerType, LongType
import yaml

import great_expectations as ge
from great_expectations.core.batch import BatchRequest, RuntimeBatchRequest
from great_expectations.data_context import BaseDataContext
from great_expectations.data_context.types.base import (
    DataContextConfig,
    InMemoryStoreBackendDefaults,
)
from great_expectations.data_context.types.base import DataContextConfig, DatasourceConfig, S3StoreBackendDefaults
from great_expectations.data_context import BaseDataContext
from great_expectations.core import RunIdentifier
from great_expectations.core.batch import BatchRequest, RuntimeBatchRequest
from great_expectations.checkpoint import SimpleCheckpoint, Checkpoint
from great_expectations.exceptions import DataContextError
from datetime import datetime
from pyspark.sql.utils import AnalysisException
from pyspark.sql import Row

str_fmt = "%m-%d-%Y"
def has_column(df, col):
    try:
        df[col]
        return True
    except AnalysisException:
        return False
```
# Configure Context
```
project_config = DataContextConfig(
    config_version=1,
    plugins_directory=None,
    config_variables_file_path=None,
    datasources={
        "my_spark_datasource": {
            "data_asset_type": {
                "class_name": "SparkDFDataset",
                "module_name": "great_expectations.dataset",
            },
            "class_name": "SparkDFDatasource",
            "module_name": "great_expectations.datasource",
            "batch_kwargs_generators": {},
        }
    },
    stores={
        "expectations_S3_store": {
            "class_name": "ExpectationsStore",
            "store_backend": {
                "class_name": "TupleS3StoreBackend",
                "bucket": "my_s3_bucket",
                "prefix": "my_expectations_store_prefix",
            },
        },
        "validations_S3_store": {
            "class_name": "ValidationsStore",
            "store_backend": {
                "class_name": "TupleS3StoreBackend",
                "bucket": "my_s3_bucket",
                "prefix": "my_validations_store_prefix",
            },
        },
        "checkpoints_S3_store": {
            "class_name": "CheckpointStore",
            "store_backend": {
                "class_name": "TupleS3StoreBackend",
                "bucket": "my_s3_bucket",
                "prefix": "my_checkpoint_store_prefix",
            },
        },
        "evaluation_parameter_store": {
            "class_name": "EvaluationParameterStore"
        },
    },
    expectations_store_name="expectations_S3_store",
    validations_store_name="validations_S3_store",
    checkpoint_store_name="validations_S3_store",
    evaluation_parameter_store_name="evaluation_parameter_store",
    data_docs_sites={
        "s3_site": {
            "class_name": "SiteBuilder",
            "store_backend": {
                "class_name": "TupleS3StoreBackend",
                "bucket": "my_s3_bucket",
                "prefix": "my_optional_data_docs_prefix",
            },
            "site_index_builder": {
                "class_name": "DefaultSiteIndexBuilder",
                "show_cta_footer": True,
            },
        }
    },
    validation_operators={
        "action_list_operator": {
            "class_name":
                "ActionListValidationOperator",
            "action_list": [
                {
                    "name": "store_validation_result",
                    "action": {
                        "class_name": "StoreValidationResultAction"
                    },
                },
                {
                    "name": "store_evaluation_params",
                    "action": {
                        "class_name": "StoreEvaluationParametersAction"
                    },
                },
                {
                    "name": "update_data_docs",
                    "action": {
                        "class_name": "UpdateDataDocsAction"
                    },
                },
            ],
        }
    },
    anonymous_usage_statistics={"enabled": True})
context = BaseDataContext(project_config=project_config)
```
# Read Spark DF
```
event = "example"
expectation_suite_name = "example_suite_name"
df = spark.read.parquet(f"s3://...")
datasource_config = {
    "name": "myDataSource",
    "class_name": "Datasource",
    "execution_engine": {
        "class_name": "SparkDFExecutionEngine"
    },
    "data_connectors": {
        "default_runtime_data_connector_name": {
            "class_name": "RuntimeDataConnector",
            "batch_identifiers": [
                "batch_id"
            ],
        }
    },
}
context.add_datasource(**datasource_config)
batch_request = RuntimeBatchRequest(
    datasource_name = "myDataSource",
    data_connector_name = "default_runtime_data_connector_name",
    data_asset_name = "myDataAsset",
    batch_identifiers = {"batch_id": "example_batch_id"},
    runtime_parameters = {"batch_data": df}
)
```
# Run Checkpoint
```
checkpoint_name = expectation_suite_name
checkpoint_config = {
    "config_version": 1.0,
    "class_name": "Checkpoint",
    "run_name_template": f"%Y-%M-{checkpoint_name}",
    "validations": [
        {
            "batch_request": batch_request,
            "expectation_suite_name": expectation_suite_name,
            "action_list": [
                {
                    'name': 'store_validation_result', 
                    'action': {
                        'class_name': 'StoreValidationResultAction'
                    }
                }, 
                {
                    'name': 'store_evaluation_params', 
                    'action': {
                        'class_name': 'StoreEvaluationParametersAction'
                    }
                }, 
                {
                    'name': 'update_data_docs', 
                    'action': {
                        'class_name': 'UpdateDataDocsAction'
                    }
                }
            ],
            "evaluation_parameters": None,
        }
    ],
}
checkpoint = Checkpoint(
    name = checkpoint_name, 
    data_context = context, 
    **checkpoint_config
)
checkpoint_result = checkpoint.run()
```
> An error was encountered:
'CheckpointConfig' object has no attribute '_get_object_id'
Traceback (most recent call last):
  File "/tmp/1633033764301-0/lib/python3.7/site-packages/great_expectations/checkpoint/checkpoint.py", line 264, in run
    runtime_kwargs=runtime_kwargs
  File "/tmp/1633033764301-0/lib/python3.7/site-packages/great_expectations/checkpoint/checkpoint.py", line 164, in get_substituted_config
    substituted_config = copy.deepcopy(config)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 281, in _reconstruct
    state = deepcopy(state, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 241, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 216, in _deepcopy_list
    append(deepcopy(a, memo))
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 241, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 241, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 241, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 281, in _reconstruct
    state = deepcopy(state, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 241, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/tmp/1633033764301-0/lib64/python3.7/copy.py", line 161, in deepcopy
    y = copier(memo)
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1248, in __call__
    args_command, temp_args = self._build_args(*args)
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1212, in _build_args
    (new_args, temp_args) = self._get_args(args)
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1199, in _get_args
    temp_arg = converter.convert(arg, self.gateway_client)
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_collections.py", line 513, in convert
    java_map[key] = object[key]
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_collections.py", line 72, in __setitem__
    self.put(key, value)
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1248, in __call__
    args_command, temp_args = self._build_args(*args)
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1218, in _build_args
    [get_command_part(arg, self.pool) for arg in new_args])
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/java_gateway.py", line 1218, in <listcomp>
    [get_command_part(arg, self.pool) for arg in new_args])
  File "/usr/lib/spark/python/lib/py4j-0.10.7-src.zip/py4j/protocol.py", line 298, in get_command_part
    command_part = REFERENCE_TYPE + parameter._get_object_id()
AttributeError: 'CheckpointConfig' object has no attribute '_get_object_id'

**Environment (please complete the following information):**
 - Operating System: AWS EMR
 - Great Expectations Version: 0.13.36

**Comment by talagluck** [comment]
Thanks for raising this issue, @tnorlund-dsc ! We will review and be in touch soon.

**Comment by tnorlund-dsc** [comment]
Thanks for guiding me over slack @bhcastleton! I'd like some guidance @talagluck and @NathanFarmer 

I previously submitted issue #3437 and #3445. I've made progress through GE's SW stack, but it seems as though that the `RuntimeBatchRequest` is still difficult to use with a `Checkpoint`. 

Is this an issue with Spark through copying the DF into the `Checkpoint`? Would you guys recommend another way of doing this? I'd just like to have a proof of concept through the data docs ASAP. 

Please let me know how I can help.

**Comment by NathanFarmer** [comment]
Hi @tnorlund-dsc! Thanks for opening this issue. I do believe there is a bug when passing the Spark df. I am working on a PR at this moment with a new test that uses Spark dfs. Having said that, I am not able to reproduce this error message. It does seem to be related based on the fact that it flows from a `deepcopy` being performed when the `Checkpoint` is run.

**Comment by NathanFarmer** [comment]
@tnorlund-dsc PR #3502 has been opened to address serialization issues when passing in a Spark DF to `Checkpoint` `validations` via `RuntimeDataConnector`. While it is under review, I will continue to try to reproduce this error message.

**Comment by tnorlund-dsc** [comment]
@NathanFarmer Thanks for letting me know what's happening. I'll have my sets of expectations ready for my datas as soon as that PR is merged. How are you running Spark? Is there a way I can repro your Spark instance so that I can get around this without the error?

**Comment by NathanFarmer** [comment]
@tnorlund-dsc PR #3502 has been merged into `develop`. It may be worth it for you to see if that resolves this issue. For Spark all I'm typically doing is something like:
```
from pyspark.sql import SparkSession
spark = SparkSession.builder.appName("Test Spark").getOrCreate()
df = spark.read.csv("/directory/file.csv", mode="DROPMALFORMED", inferSchema=True, header=True)
```

**Comment by tnorlund-dsc** [comment]
I can no longer install great expectations in EMR. I'd most likely need to bootstrap the installation. Does anyone at GE have a bootstrap script?

**Comment by NathanFarmer** [comment]
Hi @tnorlund-dsc. Is this related to this issue? It would seem to be a different issue if it is occurring when you run `sc.install_pypi_package("great_expectations")`. Is that when you see the error?

**Comment by piedpipperer** [comment]
To go further in the debugging of runtime checkpoint in spark, I tried with a simple config of chekpoint:


```
checkpoint_config = """
name: name_checkpoint_test
config_version: 1
class_name: SimpleCheckpoint
validations:
  - batch_request:
      datasource_name: spark_datasource
      data_connector_name: default_runtime_data_connector_name
      data_asset_name: name_quality_checks
      data_connector_query:
        index: -1
    expectation_suite_name: name_quality_checks
"""
check_test = context.test_yaml_config(yaml_config=checkpoint_config)
context.add_checkpoint(**yaml.load(checkpoint_config))

context.run_checkpoint(checkpoint_name="name_checkpoint_test",
    batch_request={
        "runtime_parameters": {
            "batch_data": df,
        }
     },
)
# df being a spark dataframe
```
The error "TypeError: cannot pickle '_thread.RLock' object" keeps on appearing which is complicated to debug. 
Tried before with a Checkpoint instead of SimpleCheckpoint. but guess this is simpler to debug

**Comment by NathanFarmer** [comment]
Hi @tnorlund-dsc @piedpipperer! This issue will be resolved by one of the following PRs:
- #3680 
- #3713 
- #3718 
- #3723 

The fix will be included as part of GE release 0.13.44 later this week.
